src/libcore/ffi.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation of a `va_list`.
+
+#[cfg(any(target_arch = "arm", target_arch = "mips", target_arch = "powerpc64",
+          target_arch = "x86", windows))]
+#[cfg(not(stage0))]
+/// Basic byte pointer variant of a `va_list`.
+pub type VaListImpl = *const i8;
+
+#[cfg(all(target_arch = "aarch64", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// AArch64 ABI implementation of a `va_list`. See the
+/// [Aarch64 Procedure Call Standard] for more details.
+///
+/// [AArch64 Procedure Call Standard]:
+/// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+pub struct VaListImpl {
+    stack: *mut (),
+    gr_top: *mut (),
+    vr_top: *mut (),
+    gr_offs: i32,
+    vr_offs: i32,
+}
+
+#[cfg(all(target_arch = "powerpc", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// PowerPC ABI implementation of a `va_list`.
+pub struct VaListImpl {
+    gpr: u8,
+    fpr: u8,
+    reserved: u16,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+#[cfg(all(target_arch = "x86_64", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// x86_64 ABI implementation of a `va_list`.
+struct VaListImpl {
+    gp_offset: i32,
+    fp_offset: i32,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+/// A wrapper for a `va_list`
+#[cfg(not(stage0))]
+#[lang = "va_list"]
+#[unstable(feature = "c_variadic",
+           reason = "the `c_variadic` feature has not been properly tested on \
+                     all supported platforms",
+           issue = "27745")]
+#[derive(Clone, Copy, Debug)]
+// FIXME(dlrobertson): Do not use a *mut here. Fix core::intrinsics and
+// rustc_typeck::check::intrinsic::check_intrinsic_type.
+pub struct VaList(*mut VaListImpl);
+
+#[cfg(not(stage0))]
+#[unstable(feature = "c_variadic",
+           reason = "the `c_variadic` feature has not been properly tested on \
+                     all supported platforms",
+           issue = "27745")]
+impl VaList {
+    /// Advance to the next arg.
+    pub unsafe fn arg<T: Copy>(&mut self) -> T {
+        ::intrinsics::va_arg(*self)
+    }
+
+    /// Copy the `va_list` at the current location.
+    pub unsafe fn copy<'ret, F, T>(&self, f: F) -> T
+            where F: FnOnce(VaList) -> T, T: 'ret {
```

Comment:
I don't think you need a lifetime bound on `T`, because `VaList` (if it has a lifetime parameter) can't escape the closure being called *anyway*. Due to lifetime parametricity, `F` *cannot* assume *anything* about that lifetime, and thus it can't possibly appear in its return type.

Reply:
These functions were fairly carefully defined in the RFC; I don't have any objection to changing this if it turns out to remain safe, but can we confirm that with the folks who originally contributed to these carefully constructed types? (cc @withoutboats )

Also, this seems to have removed the `for<'copy>` bound and the lifetime on the `VaList`. I'm not sure, but I think that makes it possible to return the `VaList` in ways that it shouldn't be. Or have I missed something there?

Reply:
The lifetime is temporarily missing, it's going to be back https://github.com/rust-lang/rust/pull/49878#discussion_r185693668 - as for the RFC, I guess I'm sorry it wasn't more thoroughly reviewed. It's harder *without* a prototype implementation because a lot of details feel out of context.

Reply:
I re-added the lifetime and along with it the lifetimes included in the RFC (it will currently fail to compile due to the problem I'm still trying to figure out).